async-nats/src/jetstream/stream.rs

```diff
@@ -1262,3 +1232,120 @@ pub struct External {
     #[serde(rename = "deliver", skip_serializing_if = "is_default")]
     pub delivery_prefix: Option<String>,
 }
+
+use std::marker::PhantomData;
+
+#[derive(Debug, Default)]
+pub struct Yes;
+#[derive(Debug, Default)]
+pub struct No;
+
+pub trait ToAssign: Debug {}
+
+impl ToAssign for Yes {}
+impl ToAssign for No {}
+
+#[derive(Debug)]
+pub struct PurgeFuture<SEQUENCE, KEEP>
+where
+    SEQUENCE: ToAssign,
+    KEEP: ToAssign,
+{
+    stream: crate::jetstream::stream::Stream,
+    inner: PurgeRequest,
+    sequence_set: PhantomData<SEQUENCE>,
+    keep_set: PhantomData<KEEP>,
+}
+
+impl<SEQUENCE, KEEP> PurgeFuture<SEQUENCE, KEEP>
+where
+    SEQUENCE: ToAssign,
+    KEEP: ToAssign,
+{
+    /// Adds subject filter to [PurgeRequest]
+    pub fn filter<T: Into<String>>(mut self, filter: T) -> PurgeFuture<SEQUENCE, KEEP> {
+        self.inner.filter = Some(filter.into());
+        self
+    }
+}
+
+impl PurgeFuture<No, No> {
+    pub(crate) fn build(stream: Stream) -> PurgeFuture<No, No> {
+        PurgeFuture {
+            stream,
+            inner: Default::default(),
+            sequence_set: PhantomData {},
+            keep_set: PhantomData {},
+        }
+    }
+}
+
+impl<KEEP> PurgeFuture<No, KEEP>
+where
+    KEEP: ToAssign,
+{
+    /// Creates a new [PurgeRequest].
+    /// `keep` and `sequence` are exclusive, enforced compile time by generics.
+    pub fn keep(self, keep: u64) -> PurgeFuture<No, Yes> {
+        PurgeFuture {
+            stream: self.stream,
+            sequence_set: PhantomData {},
+            keep_set: PhantomData {},
+            inner: PurgeRequest {
+                keep: Some(keep),
+                ..self.inner
+            },
+        }
+    }
+}
+impl<SEQUENCE> PurgeFuture<SEQUENCE, No>
+where
+    SEQUENCE: ToAssign,
+{
+    /// Creates a new [PurgeRequest].
+    /// `keep` and `sequence` are exclusive, enforces compile time by generics.
+    pub fn sequence(self, sequence: u64) -> PurgeFuture<Yes, No> {
+        PurgeFuture {
+            stream: self.stream,
+            sequence_set: PhantomData {},
+            keep_set: PhantomData {},
+            inner: PurgeRequest {
+                sequence: Some(sequence),
+                ..self.inner
+            },
+        }
+    }
+}
+
+impl<S, K> IntoFuture for PurgeFuture<S, K>
+where
+    S: ToAssign + std::marker::Send,
+    K: ToAssign + std::marker::Send,
+{
+    type Output = Result<PurgeResponse, Error>;
+
+    type IntoFuture = Pin<Box<dyn Future<Output = Result<PurgeResponse, Error>> + Send>>;
+
+    fn into_future(self) -> Self::IntoFuture {
+        Box::pin(std::future::IntoFuture::into_future(send_purge_request(
+            self.stream,
+            self.inner,
+        )))
+    }
+}
+async fn send_purge_request(stream: Stream, request: PurgeRequest) -> Result<PurgeResponse, Error> {
```

Comment:
Inline this inside into_future?

```suggestion
async fn send_purge_request(stream: Stream, request: PurgeRequest) -> Result<PurgeResponse, Error> {
```